File: archive/python-legacy/meept/skills/parser.py
"""SKILL.md parser -- YAML frontmatter + Markdown body.

Parses skill definition files in the OpenCode-style ``SKILL.md`` format:
a YAML frontmatter block (delimited by ``---``) followed by a Markdown body
containing instructions and examples.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path

import yaml

log = logging.getLogger(__name__)


@dataclass(slots=True)
class SkillMetadata:
    """Structured metadata extracted from the YAML frontmatter."""

    name: str = ""
    description: str = ""
    requires: list[str] = field(default_factory=list)
    allowed_tools: list[str] = field(default_factory=list)
    risk_level: str = "medium"
    max_iterations: int = 10
    temperature: float | None = None
    max_tokens: int | None = None


@dataclass(slots=True)
class ParsedSkill:
    """A fully parsed SKILL.md file: metadata + Markdown body."""

    metadata: SkillMetadata
    body: str
    source_path: Path | None = None


def parse_skill_file(path: Path) -> ParsedSkill | None:
    """Parse a ``SKILL.md`` file into a :class:`ParsedSkill`.

    Returns ``None`` if the file cannot be parsed or has no ``name``.
    """
    try:
        text = path.read_text(encoding="utf-8")
    except OSError:
        log.warning("Could not read skill file: %s", path)
        return None

    return parse_skill_text(text, source_path=path)


def parse_skill_text(text: str, source_path: Path | None = None) -> ParsedSkill | None:
    """Parse raw SKILL.md text into a :class:`ParsedSkill`.

    Returns ``None`` if the frontmatter is invalid or has no ``name``.
    """
    frontmatter, body = _split_frontmatter(text)
    if frontmatter is None:
        log.warning("No YAML frontmatter found in skill file: %s", source_path)
        return None

    try:
        data = yaml.safe_load(frontmatter)
    except yaml.YAMLError:
        log.warning("Invalid YAML frontmatter in skill file: %s", source_path, exc_info=True)
        return None

    if not isinstance(data, dict):
        log.warning("Frontmatter is not a mapping in skill file: %s", source_path)
        return None

    name = data.get("name", "")
    if not name:
        log.warning("Skill file has no 'name' in frontmatter: %s", source_path)
        return None

    requires_raw = data.get("requires", [])
    if isinstance(requires_raw, str):
        requires_raw = [r.strip() for r in requires_raw.split(",")]

    allowed_raw = data.get("allowed-tools", data.get("allowed_tools", []))
    if isinstance(allowed_raw, str):
        allowed_raw = [t.strip() for t in allowed_raw.split(",")]

    metadata = SkillMetadata(
        name=str(name),
        description=str(data.get("description", "")),
        requires=list(requires_raw),
        allowed_tools=list(allowed_raw),
        risk_level=str(data.get("risk-level", data.get("risk_level", "medium"))),
        max_iterations=int(data.get("max-iterations", data.get("max_iterations", 10))),
        temperature=data.get("temperature"),
        max_tokens=data.get("max-tokens", data.get("max_tokens")),
    )

    return ParsedSkill(
        metadata=metadata,
        body=body.strip(),
        source_path=source_path,
    )


def _split_frontmatter(text: str) -> tuple[str | None, str]:
    """Split ``---``-delimited YAML frontmatter from the Markdown body.

    Returns ``(frontmatter_text, body_text)``.  If no valid frontmatter
    delimiters are found, returns ``(None, original_text)``.
    """
    stripped = text.lstrip()
    if not stripped.startswith("---"):
        return None, text

    # Find the closing ---
    first_break = stripped.index("---")
    rest = stripped[first_break + 3:]
    newline_pos = rest.find("\n")
    if newline_pos == -1:
        return None, text

    rest = rest[newline_pos + 1:]

    close_pos = rest.find("\n---")
    if close_pos == -1:
        # Try end-of-string ---
        if rest.rstrip().endswith("---"):
            close_pos = rest.rstrip().rfind("---")
            frontmatter = rest[:close_pos]
            body = ""
        else:
            return None, text
    else:
        frontmatter = rest[:close_pos]
        body = rest[close_pos + 4:]

    return frontmatter, body

File: internal/skills/parser.go
package skills

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// Parser errors.
var (
	ErrNoFrontmatter = errors.New("no YAML frontmatter found")
	ErrInvalidYAML   = errors.New("invalid YAML frontmatter")
	ErrNoName        = errors.New("skill has no name in frontmatter")
)

// ParseError wraps a parsing error with file path context.
type ParseError struct {
	Path    string
	Message string
	Cause   error
}

func (e *ParseError) Error() string {
	if e.Cause != nil {
		return e.Path + ": " + e.Message + ": " + e.Cause.Error()
	}
	return e.Path + ": " + e.Message
}

func (e *ParseError) Unwrap() error {
	return e.Cause
}

// ParseSkillMetadataOnly parses only the YAML frontmatter from a SKILL.md file.
// This is faster than ParseSkillFile as it skips parsing the body content.
// YAML frontmatter is optional: when absent the slug (directory name) is used as the name.
func ParseSkillMetadataOnly(path string) (*SkillIndexEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, &ParseError{
			Path:    path,
			Message: "failed to read file",
			Cause:   err,
		}
	}

	frontmatter, _, err := splitFrontmatter(string(data))
	if err != nil {
		if errors.Is(err, ErrNoFrontmatter) {
			// No frontmatter — derive name from directory.
			// Return valid entry alongside ErrNoFrontmatter so callers can warn.
			slug := slugFromPath(path)
			return &SkillIndexEntry{
				Name:      slug,
				Path:      path,
				RiskLevel: "medium",
			}, ErrNoFrontmatter
		}
		return nil, &ParseError{
			Path:    path,
			Message: "failed to split frontmatter",
			Cause:   err,
		}
	}

	meta, err := parseMetadata(frontmatter)
	if err != nil {
		return nil, &ParseError{
			Path:    path,
			Message: "failed to parse metadata",
			Cause:   err,
		}
	}

	if meta.Name == "" {
		return nil, &ParseError{
			Path:    path,
			Message: "skill has no name",
			Cause:   ErrNoName,
		}
	}

	entry := &SkillIndexEntry{
		Name:         meta.Name,
		Description:  meta.Description,
		Requires:     meta.Requires,
		Tags:         meta.Tags,
		Path:         path,
		RiskLevel:    meta.RiskLevel,
		AllowedTools: meta.AllowedTools,
		Examples:     meta.Examples,
	}

	// Apply defaults
	if entry.RiskLevel == "" {
		entry.RiskLevel = "medium"
	}

	return entry, nil
}

// ParseSkillFile parses a SKILL.md file and returns a Skill.
// Returns an error if the file cannot be read or has invalid format.
func ParseSkillFile(path string) (*Skill, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, &ParseError{
			Path:    path,
			Message: "failed to read file",
			Cause:   err,
		}
	}

	skill, err := ParseSkillText(string(data))
	if err != nil {
		return nil, &ParseError{
			Path:    path,
			Message: "failed to parse",
			Cause:   err,
		}
	}

	skill.Path = path
	return skill, nil
}

// ParseSkillText parses raw SKILL.md text and returns a Skill.
// YAML frontmatter is optional: when absent the full text becomes the body.
func ParseSkillText(text string) (*Skill, error) {
	frontmatter, body, err := splitFrontmatter(text)
	if err != nil {
		if errors.Is(err, ErrNoFrontmatter) {
			// No frontmatter — use full text as body, no name set.
			// Return valid skill alongside ErrNoFrontmatter so callers can warn.
			return &Skill{
				Body:          strings.TrimSpace(text),
				RiskLevel:     "medium",
				MaxIterations: 10,
			}, ErrNoFrontmatter
		}
		return nil, err
	}

	meta, err := parseMetadata(frontmatter)
	if err != nil {
		return nil, err
	}

	if meta.Name == "" {
		return nil, ErrNoName
	}

	skill := &Skill{
		Name:          meta.Name,
		Description:   meta.Description,
		Requires:      meta.Requires,
		Tags:          meta.Tags,
		Examples:      meta.Examples,
		Body:          strings.TrimSpace(body),
		AllowedTools:  meta.AllowedTools,
		RiskLevel:     meta.RiskLevel,
		MaxIterations: meta.MaxIterations,
		Temperature:   meta.Temperature,
		MaxTokens:     meta.MaxTokens,
		MCPServers:    meta.MCPServers,
	}

	// Apply defaults
	if skill.RiskLevel == "" {
		skill.RiskLevel = "medium"
	}
	if skill.MaxIterations == 0 {
		skill.MaxIterations = 10
	}

	return skill, nil
}

// splitFrontmatter splits YAML frontmatter from the markdown body.
// The frontmatter must be delimited by --- markers.
func splitFrontmatter(text string) (frontmatter, body string, err error) {
	trimmed := strings.TrimLeft(text, " \t\n\r")
	if !strings.HasPrefix(trimmed, "---") {
		return "", "", ErrNoFrontmatter
	}

	// Find the opening ---
	_, after, ok := strings.Cut(trimmed, "---")
	if !ok {
		return "", "", ErrNoFrontmatter
	}

	// Skip past the opening marker and any trailing content on that line
	rest := after
	newlinePos := strings.Index(rest, "\n")
	if newlinePos == -1 {
		return "", "", ErrNoFrontmatter
	}
	rest = rest[newlinePos+1:]

	// Find the closing --- (can be at start of a line or immediately after opening)
	// Check for --- at start of remaining content (empty frontmatter case)
	if strings.HasPrefix(rest, "---") {
		// Empty frontmatter
		afterClose := rest[3:]
		_, after, ok := strings.Cut(afterClose, "\n")
		if ok {
			body = after
		} else {
			body = strings.TrimPrefix(afterClose, "\n")
		}
		return "", body, nil
	}

	// Find the closing --- on its own line
	closePos := strings.Index(rest, "\n---")
	if closePos == -1 {
		// Try end-of-string ---
		trimmedRest := strings.TrimRight(rest, " \t\n\r")
		if strings.HasSuffix(trimmedRest, "---") {
			closePos = strings.LastIndex(trimmedRest, "---")
			frontmatter = rest[:closePos]
			body = ""
			return frontmatter, body, nil
		}
		return "", "", ErrNoFrontmatter
	}

	frontmatter = rest[:closePos]
	// Skip the \n--- and any content after the closing marker line
	afterClose := rest[closePos+4:]
	_, after0, ok0 := strings.Cut(afterClose, "\n")
	if ok0 {
		body = after0
	} else {
		body = ""
	}

	return frontmatter, body, nil
}

// parseMetadata parses the YAML frontmatter into SkillMetadata.
func parseMetadata(frontmatter string) (*SkillMetadata, error) {
	// Start with zero values to detect which fields were actually set
	var meta SkillMetadata

	if err := yaml.Unmarshal([]byte(frontmatter), &meta); err != nil {
		return nil, ErrInvalidYAML
	}

	// Handle alternative field names (with underscores instead of hyphens).
	// The frontmatter was already validated above, so this parse cannot fail.
	var altMeta struct {
		AllowedTools  []string `yaml:"allowed_tools"`
		RiskLevel     string   `yaml:"risk_level"`
		MaxIterations int      `yaml:"max_iterations"`
		MaxTokens     *int     `yaml:"max_tokens"`
	}
	if err := yaml.Unmarshal([]byte(frontmatter), &altMeta); err != nil {
		return nil, fmt.Errorf("parse alt skill metadata: %w", err)
	}

	// Merge alternative field values if primary is empty/zero
	if len(meta.AllowedTools) == 0 && len(altMeta.AllowedTools) > 0 {
		meta.AllowedTools = altMeta.AllowedTools
	}
	if meta.RiskLevel == "" && altMeta.RiskLevel != "" {
		meta.RiskLevel = altMeta.RiskLevel
	}
	if meta.MaxIterations == 0 && altMeta.MaxIterations != 0 {
		meta.MaxIterations = altMeta.MaxIterations
	}
	if meta.MaxTokens == nil && altMeta.MaxTokens != nil {
		meta.MaxTokens = altMeta.MaxTokens
	}

	return &meta, nil
}

// slugFromPath derives a skill slug from the file path by using the parent directory name.
func slugFromPath(path string) string {
	dir := filepath.Dir(path)
	if dir != "" && dir != "." {
		return filepath.Base(dir)
	}
	return "unknown"
}
